### PR TITLE
fix: Add labels to ephemeral (listener) volumes

### PR DESCRIPTION
**Pull request by razvan**
# Description

:green_circle:  CI https://ci.stackable.tech/view/02%20Operator%20Tests%20(custom)/job/hdfs-operator-it-custom/143/

Part of https://github.com/stackabletech/issues/issues/555

Depends on https://github.com/stackabletech/operator-rs/pull/799

Ensure the ephemeral (listener) volumes attached directly to the datanode containers are labeled with the "recommended labels".

This is now the expected structure:

```
metadata:
  name: hdfs-datanode-default
spec:
  template:
    volumes:
    - ephemeral:
        volumeClaimTemplate:
          metadata:
            annotations:
              listeners.stackable.tech/listener-class: external-unstable
            creationTimestamp: null
            labels:
              app.kubernetes.io/component: datanode
              app.kubernetes.io/instance: hdfs
              app.kubernetes.io/name: hdfs
              app.kubernetes.io/role-group: default
          spec:
            accessModes:
            - ReadWriteMany
            resources:
              requests:
                storage: "1"
            storageClassName: listeners.stackable.tech
            volumeMode: Filesystem
      name: listener
```

Stacklet listing now includes datanode endpoints as they inherit the labels of the (listener) ephemeral volumes:

```
❯ stackablectl stacklet list -n kuttl-test-direct-wolf

┌───────────┬─────────┬────────────────────────┬──────────────────────────────────────────────────────────────┬─────────────────────────────────┐
│ PRODUCT   ┆ NAME    ┆ NAMESPACE              ┆ ENDPOINTS                                                    ┆ CONDITIONS                      │
╞═══════════╪═════════╪════════════════════════╪══════════════════════════════════════════════════════════════╪═════════════════════════════════╡
│ hdfs      ┆ hdfs    ┆ kuttl-test-direct-wolf ┆ datanode-default-0-listener-data     172.18.0.2:30580        ┆                                 │
│           ┆         ┆                        ┆ datanode-default-0-listener-http     http://172.18.0.2:32297 ┆                                 │
│           ┆         ┆                        ┆ datanode-default-0-listener-ipc      172.18.0.2:31998        ┆                                 │
│           ┆         ┆                        ┆ datanode-default-0-listener-metrics  172.18.0.2:30805        ┆                                 │
│           ┆         ┆                        ┆ datanode-default-1-listener-data     172.18.0.2:30990        ┆                                 │
│           ┆         ┆                        ┆ datanode-default-1-listener-http     http://172.18.0.2:32605 ┆                                 │
│           ┆         ┆                        ┆ datanode-default-1-listener-ipc      172.18.0.2:30825        ┆                                 │
│           ┆         ┆                        ┆ datanode-default-1-listener-metrics  172.18.0.2:30264        ┆                                 │
│           ┆         ┆                        ┆ namenode-default-0-http              http://172.18.0.2:31438 ┆                                 │
│           ┆         ┆                        ┆ namenode-default-0-metrics           172.18.0.2:31633        ┆                                 │
│           ┆         ┆                        ┆ namenode-default-0-rpc               172.18.0.2:31447        ┆                                 │
│           ┆         ┆                        ┆ namenode-default-1-http              http://172.18.0.2:30285 ┆                                 │
│           ┆         ┆                        ┆ namenode-default-1-metrics           172.18.0.2:32287        ┆                                 │
│           ┆         ┆                        ┆ namenode-default-1-rpc               172.18.0.2:30711        ┆                                 │
├╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤
│ zookeeper ┆ hdfs-zk ┆ kuttl-test-direct-wolf ┆ zk                                   172.18.0.2:32241        ┆ Available, Reconciling, Running │
└───────────┴─────────┴────────────────────────┴──────────────────────────────────────────────────────────────┴─────────────────────────────────┘

Use "stackablectl stacklet credentials [OPTIONS] <PRODUCT_NAME> <STACKLET_NAME>" to display credentials for deployed stacklets.

```

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] CRD documentation for all fields, following the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs-style-guide).
- [ ] Helm chart can be installed and deployed operator works
- [ ] Integration tests passed (for non trivial changes)
- [ ] Changes need to be "offline" compatible
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] Code contains useful logging statements
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated. Follows the [style guide](https://docs.stackable.tech/home/nightly/contributor/docs-style-guide).
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
- [ ] [Roadmap](https://github.com/orgs/stackabletech/projects/25/views/1) has been updated
```

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -28,5 +28,6 @@ tokio = { version = "1.37", features = ["full"] }
 tracing = "0.1"
 tracing-futures = { version = "0.2", features = ["futures-03"] }
 
-# [patch."https://github.com/stackabletech/operator-rs.git"]
-# stackable-operator = { git = "https://github.com/stackabletech//operator-rs.git", branch = "main" }
+[patch."https://github.com/stackabletech/operator-rs.git"]
+#stackable-operator = { path = "../operator-rs/crates/stackable-operator" }
+stackable-operator = { git = "https://github.com/stackabletech//operator-rs.git", branch = "bugfix/pvc-labels" }
```
Comment (sbernauer) [inline Cargo.toml]: Reminder to use released version

File: rust/operator-binary/src/hdfs_controller.rs [hunk 2]
```diff
@@ -349,22 +349,43 @@ pub async fn reconcile_hdfs(hdfs: Arc<HdfsCluster>, ctx: Arc<Ctx>) -> HdfsOperat
 
             let rolegroup_ref = hdfs.rolegroup_ref(role_name, rolegroup_name);
 
-            let rg_service =
-                rolegroup_service(&hdfs, &role, &rolegroup_ref, &resolved_product_image)?;
+            // The "met_builder" variable is needed to keep the rust compiler happy.
+            // Without it, an error is thrown that the "metadata" variable is dropped
+            // while still being in use.
+            let mut met_builder = ObjectMetaBuilder::new();
+            let metadata = met_builder
```
Comment (sbernauer) [inline rust/operator-binary/src/hdfs_controller.rs]: Thanks for the comment! First off there is a typo, but I personally have the feeling this is pretty normal stuff and the documentation might mislead readers that this situation is different (I have to admit I didn't understood it).

We can also shadow `metadata` here

```suggestion
            let mut metadata = ObjectMetaBuilder::new();
            let metadata = metadata
```

**Review by sbernauer — CHANGES_REQUESTED** [review]
LGTM otherwise

**Comment by sbernauer** [comment]
~I noticed the journalnodes are missing Listeners. As you can **not** port-forward, there is no way to access the journalnode UI AFAIK. I think we should expose that UI as well, what do you think?~

Was already discussed in [Slack](https://stackable-workspace.slack.com/archives/C02FZ581UCD/p1705573773179029?thread_ts=1704714818.413279&cid=C02FZ581UCD), no need to expose the JN UI

File: rust/operator-binary/src/hdfs_controller.rs (hunk 2 above)
Comment (razvan) [reply rust/operator-binary/src/hdfs_controller.rs]: I don't understand the behavior either but if I remove line 352 and change line 353 with:
```
            let mut metadata = ObjectMetaBuilder::new()
```
you'll get
```
error[E0716]: temporary value dropped while borrowed
   --> rust/operator-binary/src/hdfs_controller.rs:356:32
    |
356 |             let mut metadata = ObjectMetaBuilder::new()
    |                                ^^^^^^^^^^^^^^^^^^^^^^^^ creates a temporary value which is freed while still in use
...
370 |                 .context(ObjectMetaSnafu)?;
    |                                           - temporary value is freed at the end of this statement
371 |
372 |             let rg_service = rolegroup_service(&hdfs, metadata, &role, &rolegroup_ref)?;
    |                                                       -------- borrow later used here
    |
    = note: consider using a `let` binding to create a longer lived value
```

Comment (razvan) [reply rust/operator-binary/src/hdfs_controller.rs]: I didn't apply your suggestion, instead I tried to clarify the comment too: https://github.com/stackabletech/hdfs-operator/pull/534/commits/bb2a85f137f66c3cef2ed2c1e3092854ba41fa84

Comment (sbernauer) [reply rust/operator-binary/src/hdfs_controller.rs]: Perfectly fine, thanks!

File: Cargo.toml [hunk 3]
```diff
@@ -28,5 +28,6 @@ tokio = { version = "1.37", features = ["full"] }
 tracing = "0.1"
 tracing-futures = { version = "0.2", features = ["futures-03"] }
 
-# [patch."https://github.com/stackabletech/operator-rs.git"]
-# stackable-operator = { git = "https://github.com/stackabletech//operator-rs.git", branch = "main" }
+[patch."https://github.com/stackabletech/operator-rs.git"]
+#stackable-operator = { path = "../operator-rs/crates/stackable-operator" }
+stackable-operator = { git = "https://github.com/stackabletech//operator-rs.git", branch = "main" }
```
Comment (sbernauer) [inline Cargo.toml]: [stackable-operator-0.69.0](https://github.com/stackabletech/operator-rs/releases/tag/stackable-operator-0.69.0) is out :)

Comment (razvan) [reply Cargo.toml]: thanks https://github.com/stackabletech/hdfs-operator/pull/534/commits/924f49b3baf0ac79be67604bb808f83b7b789863

**Review by sbernauer — APPROVED** [review]

**Pull request merged**